src/components/character-controller.js

```diff
@@ -88,11 +86,29 @@ AFRAME.registerComponent("character-controller", {
     this.el.emit("snap_rotate_right");
   },
 
-  handleTeleport: function(event) {
-    const position = event.detail.newPosition;
-    const navPosition = event.detail.hitPoint;
-    this.resetPositionOnNavMesh(position, navPosition, this.el.object3D);
-  },
+  // We assume the rig is at the root, and its local position === its world position.
+  teleportTo: (function() {
+    const rig = new THREE.Vector3();
+    const head = new THREE.Vector3();
+    const deltaFromHeadToTarget = new THREE.Vector3();
+    const targetForRig = new THREE.Vector3();
+    return function teleportTo(targetWorldPosition) {
+      const o = this.el.object3D;
+      o.getWorldPosition(rig);
+      this.data.pivot.object3D.getWorldPosition(head);
+      deltaFromHeadToTarget.copy(targetWorldPosition).sub(head);
+      targetForRig.copy(rig).add(deltaFromHeadToTarget);
+
+      const pathfinder = this.el.sceneEl.systems.nav.pathfinder;
+      this.navGroup = pathfinder.getGroup(NAV_ZONE, targetForRig, true, true);
+      this.navNode = null;
+      this._setNavNode(targetForRig);
+      pathfinder.clampStep(rig, targetForRig, this.navNode, NAV_ZONE, this.navGroup, o.position);
+      o.matrixNeedsUpdate = true;
+      o.updateMatrices();
```

Comment:
not sure if I should call updateMatrices and decompose here or if i should do it in tick instead, since that's where we need this to be up to date.

Reply:
You are setting position here, so the only thing you need to do is set matrixNeedsUpdate. In tick where you area actually reading `root.matrix` is where `updateMatricies` needs to get called. I don't think a decompose is needed until later in tick where it is already being done ` root.matrix.decompose(root.position, root.quaternion, root.scale);`. Here its a noop since yo literally just set vector3 position, then updated the matrix with that position, then decomposed the position you just set back into the vector3 position.